src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs

```diff
@@ -302,16 +302,10 @@ internal void Cancel(MySqlCommand command)
 			if (m_session.ServerVersion.Version < ServerVersions.SupportsProcedureCache)
 				return null;
 
-			if (m_cachedProcedures == null)
-				m_cachedProcedures = new Dictionary<string, CachedProcedure>();
-
 			var normalized = NormalizedSchema.MustNormalize(name, Database);
-			if (!m_cachedProcedures.TryGetValue(normalized.FullyQualified, out var cachedProcedure))
-			{
-				cachedProcedure = await CachedProcedure.FillAsync(ioBehavior, this, normalized.Schema, normalized.Component, cancellationToken).ConfigureAwait(false);
-				m_cachedProcedures[normalized.FullyQualified] = cachedProcedure;
-			}
-			return cachedProcedure;
+
+			return await m_cachedProcedures.GetOrAdd(normalized.FullyQualified,
```

Comment:
I suspect this is unsafe because `normalized.FullyQualified` isn't globally unique across all servers and databases that an arbitrary .NET process might connect to.